### ISSUE TITLE
docstrings

### ISSUE THREAD

**Issue by agzam** [bakpakin/Fennel#154]
Would it be possible to add support for proper docstrings? Or even cooler - metadata in general.

**Comment by jaawerth** [comment]
It is! I've got it mostly implemented [in this fork](https://github.com/jaawerth/Fennel/tree/add-metadata-and-doc), which works by keeping a weakly referenced table of metadata that uses the function itself as a key, and once in there the same table can eventually be used for other metadata in addition to docstrings.

The only thing left to do (besides bringing that branch up to date with master since it's fallen behind at this point) is propagating the relevant compiler option to every part of the compiler that needs it (see below). I've been kinda tied up, but should be able to make time for it in a couple weeks. However, I've kind of let it drag, so I won't object if someonewants to take over finishing it and submitting the PR so they don't have to wait :wink: 

## Last step: option propagation

Since docstrings and metadata introduce a nonzero amount of overhead, they need to be disabled by default in compiled code, enabled by default in the REPL, and have an option in the compiler API/CLI to enable/disable it. This is somewhat complicated by the fact that the compiler options are manually spliced apart and passed piecemeal in various parts of the compiler code - so before this is done, I (or whoever) just need to go through and find all of those places where the metadata option needs to be propagated so that it will continue to work within nested expressions, eval, etc.

It shouldn't actually take _too_ long to do once all the places the options need to be passed on are mapped out (which will also require modifying some of the inner compiler functions so they can take an options arg).

**Comment by agzam** [comment]
Whoa... that's so awesome. Thank you for your work @jaawerth

**Comment by jaawerth** [comment]
Update: while updating my [fork](https://github.com/jaawerth/Fennel/tree/add-metadata-and-doc) on the newest master I noticed the new `rootOptions` for referring to top-level options to the compiler and switched over to using that. From the REPL and when using --compile it now appears to work correctly for all the nested compilation variants I've tried so far. Last step is to make sure the `useMetadata` is passed into/referred to by `eval` and possibly `loadCode`, add some tests/documentation, and this should be good to go!
